sdks/go/pkg/beam/core/runtime/exec/sdf.go

```diff
@@ -343,7 +348,12 @@ func (n *ProcessSizedElementsAndRestrictions) ProcessElement(_ context.Context,
 		defer func() {
 			<-n.SU
 		}()
-		return n.PDo.processSingleWindow(mainIn)
+		continuation, processResult := n.PDo.processSingleWindow(mainIn)
+		if continuation != nil {
+			n.source.pc = continuation
+			n.source.selfSu = n
```

Comment:
I think that works up at the DataSource level if we wanted to (that's how we init the SplittableUnit.) That's cleaner with how I was thinking about doing the actual split, too (trying to keep all of the logic within DataSource and just returning the split to the harness if we do so.) I'll mock up the code for that after I go through the rest of the comments

Reply:
Changed, PTAL

Reply:
FYI I do like leveraging the existing "ownership" mechanism using the channel as the changed approach seems to be taking.
As long as we clearly consider who is writing, and reading when it's fine.  We want to avoid heavy mutexes on a per element basis if we can manage it (that can get very expensive very quickly).

Reply:
The channel isn't being used if I understand this comment correctly, that was an approach discussed in the design doc that was rejected

Reply:
You've read it more recently than I have.
We can see if it's required in an independant PR to vet that and possibly get rid of it. Do Not 
combine it with the current work. That's how mistakes happen.